modules/DNN/Perceptron.cpp

```diff
@@ -42,16 +42,24 @@ namespace DirectGraphicalModels {
 		{
 			Mat error = gt - solution;
 			for (int i = 0; i < error.rows; i++)
-				error.at<float>(i, 0) *= m_vpNeuronLayers[2]->getActivationFunctionDeriateve()(solution.at<float>(i, 0));
+				error.at<float>(i, 0) *= m_vpNeuronLayers.back()->getActivationFunctionDeriateve()(solution.at<float>(i, 0));
+
+			int numLayers = m_vpNeuronLayers.size() - 1; // number of layers [0, 1, 2 ... n]
+			int numHiddenLayers = numLayers - 1;
 			
-			Mat DeltaIn_j; // = layerC.getWeights() x error;
-			gemm(m_vpNeuronLayers[2]->getWeights(), error, 1, Mat(), 0, DeltaIn_j);
+			std::vector<Mat> Error; //Vector containing errors for each of the layers neurons
+			Error.push_back(error);
 
-			// layerC.m_weights += learningRate * layerB.m_values x error.t()
-			gemm(m_vpNeuronLayers[1]->getValues(), error.t(), learningRate, m_vpNeuronLayers[2]->getWeights(), 1, m_vpNeuronLayers[2]->getWeights());
+			for (int i = 0; i < numHiddenLayers; i++)
+			{
+				Mat temp = m_vpNeuronLayers[numLayers - i]->getWeights() * Error[i]; // weights * error
```

Comment:
Well, now I think that here the derivatives of the activations functions must participate...
Let me check...